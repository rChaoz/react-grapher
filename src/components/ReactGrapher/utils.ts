import {errorCustom, errorParsingDOMElement, errorUnknownEdge, errorUnknownNode} from "../../util/log";
import {Node, NodeConfig, NodeHandleConfig, NodeHandleInfo, NodeImpl, Nodes, NodesImpl} from "../../data/Node";
import {Edge, EdgeConfig, Edges, EdgesImpl} from "../../data/Edge";
import {GrapherChange} from "../../data/GrapherChange";
import {Controller} from "../../data/Controller";
import {GrapherConfigSet, GrapherFitViewConfigSet, GrapherViewportControlsSet} from "../../data/GrapherConfig";
import {convertToCSSLength, resolveValue} from "../../util/utils";
import {CommonGraphProps} from "./props";
import {GrapherEvent} from "../../data/GrapherEvent";

const whitespaceRegex = /^\s+$/
const allowedConnectionsRegex = /^([a-zA-Z0-9_-]+)\s*(<->|->|<-)\s*([a-zA-Z0-9_-]+)/

type AllowedConnections = Map<string, string[]> & { sources: Set<string>, targets: Set<string> }

/**
 * Parse allowed connections config option
 */
export function parseAllowedConnections(s: string): AllowedConnections {
    const map: AllowedConnections = Object.assign(new Map<string, string[]>(), {sources: new Set<string>(), targets: new Set<string>()})
    // Parse config option
    let index = 0
    while (index < s.length) {
        if (!whitespaceRegex.test(s[index])) {
            const match = s.substring(index).match(allowedConnectionsRegex)
            if (match == null) {
                errorCustom(`Error while parsing config option 'allowedEdges' at index ${index}:\n`, s)
                return Object.assign(new Map<string, string[]>(), {sources: new Set<string>(), targets: new Set<string>()})
            }
            const first = match[1], second = match[3], direction = match[2]

            if (direction === "->") {
                if (!map.has(first)) map.set(first, [])
                map.get(first)!.push(second)
                map.sources.add(first)
                map.targets.add(second)
            } else if (direction === "<-") {
                if (!map.has(second)) map.set(second, [])
                map.get(second)!.push(first)
                map.sources.add(second)
                map.targets.add(first)
            } else {
                if (!map.has(first)) map.set(first, [])
                if (!map.has(second)) map.set(second, [])
                map.get(first)!.push(second)
                map.sources.add(first)
                map.targets.add(second)
                map.get(second)!.push(first)
                map.sources.add(second)
                map.targets.add(first)
            }

            index += match[0].length
        } else ++index
    }
    return map
}

/**
 * Extract DOM ID, type (node/edge) and internal ID from event target
 */
export function processDomElement<N, E>(element: EventTarget | null, nodes: Nodes<N>, edges: Edges<E>):
    { type: "node", objID: string, obj: Node<N> } | { type: "edge", objID: string, obj: Edge<E> } | { type: "handle", objID: string, obj: NodeHandleInfo, node: Node<N> } | null {
    if (element == null) {
        errorCustom("Node/Edge/Handle pointer event listener callback was called with a null event.currentTarget")
        return null
    }
    const elem = element as HTMLElement
    const type = elem.dataset.type, objID = elem.dataset.id

    if (type == null || objID == null) {
        errorParsingDOMElement(elem)
        errorCustom("Element does not have 'data-type' or 'data-id' attributes.")
        return null
    }

    let obj, nodeID, node
    switch (type) {
        case "node":
            obj = nodes.get(objID)
            if (obj == null) {
                errorParsingDOMElement(elem)
                errorUnknownNode(objID)
                return null
            }
            return {type: "node", objID, obj}
        case "edge":
            obj = edges.get(objID)
            if (obj == null) {
                errorParsingDOMElement(elem)
                errorUnknownEdge(objID)
                return null
            }
            return {type: "edge", objID, obj}
        case "handle":
            nodeID = elem.dataset.node
            if (nodeID == null) {
                errorParsingDOMElement(elem)
                errorCustom("Handle element does not have 'data-node' attribute.")
                return null
            }
            node = nodes.get(nodeID)
            if (node == null) {
                errorParsingDOMElement(elem)
                errorUnknownNode(nodeID)
                return null
            }
            obj = (node as NodeImpl<N>).handles.find(h => h.name === objID)
            if (obj == null) {
                errorParsingDOMElement(elem)
                errorCustom(`Node '${nodeID}' does not have a handle named '${objID}'`)
                return null
            }
            return {type: "handle", objID, obj, node}
    }

    errorParsingDOMElement(elem)
    return null
}

/**
 * Send a change to the graph
 */
export function sendChanges<N, E>(changes: GrapherChange[], d: { onChange: CommonGraphProps["onChange"], nodes: NodesImpl<N>, edges: EdgesImpl<E> }) {
    let c: GrapherChange[] | undefined | void = changes
    if (d.onChange != null) c = d.onChange(changes)
    if (c != null) {
        d.nodes.processChanges(c)
        d.edges.processChanges(c)
    }
}

/**
 * Send an event to the graph
 */
export function sendEvent<N, E>(event: GrapherEvent, d: { onEvent: CommonGraphProps["onEvent"], onChange: CommonGraphProps["onChange"], nodes: NodesImpl<N>, edges: EdgesImpl<E> }) {
    if (d.onEvent == null) return
    const c = d.onEvent(event)
    if (c != null) sendChanges(c, d)
}

/**
 * Zoom the viewport
 */
export function changeZoom(amount: number, controller: Controller, config: GrapherConfigSet) {
    const viewport = controller.getViewport()
    const zoom = viewport.zoom * (1 + amount)
    controller.updateViewport({
        zoom: Math.min(Math.max(zoom, config.viewportControls.minZoom), config.viewportControls.maxZoom)
    })
}

/**
 * Fit the viewport
 */
export function fitView(fitConfig: GrapherFitViewConfigSet, viewportControls: GrapherViewportControlsSet, controller: Controller, boundingRect: DOMRect, element: HTMLElement) {
    if (element == null || boundingRect == null || (boundingRect.width === 0 && boundingRect.height === 0)) return
    const w = element.offsetWidth, h = element.offsetHeight

    // Calculate zoom value for paddings
    const rect = new DOMRect(boundingRect.x, boundingRect.y, boundingRect.width, boundingRect.height)
    let zoom = Math.min(w / rect.width, h / rect.height)
    if (fitConfig.abideMinMaxZoom) zoom = Math.min(Math.max(zoom, viewportControls.minZoom), viewportControls.maxZoom)

    // Apply padding
    element.style.padding = convertToCSSLength(fitConfig.padding)
    const comp = getComputedStyle(element)
    const pl = resolveValue(comp.paddingLeft, w) / zoom, pt = resolveValue(comp.paddingTop, h) / zoom,
        pr = resolveValue(comp.paddingRight, w) / zoom, pb = resolveValue(comp.paddingBottom, h) / zoom
    element.style.padding = "0"
    rect.x -= pl
    rect.y -= pt
    rect.width += pl + pr
    rect.height += pt + pb
    // Calculate final zoom value
    zoom = Math.min(w / rect.width, h / rect.height)
    if (fitConfig.abideMinMaxZoom) zoom = Math.min(Math.max(zoom, viewportControls.minZoom), viewportControls.maxZoom)

    // Update viewport
    controller.setViewport({
        centerX: (rect.left + rect.right) / 2, centerY: (rect.top + rect.bottom) / 2, zoom
    })
}

const globalNodeDefaults: Required<NodeConfig> = {
    // Node
    allowGrabbing: true,
    allowSelection: true,
    allowMoving: true,
    allowDeletion: false,
    allowNewEdges: false,
    allowNewEdgeTarget: false,
    // Handles
    allowGrabbingHandles: false,
    allowNewEdgesFromHandles: false,
    allowNewEdgeTargetForHandles: false,
}

const globalEdgeDefaults: Required<EdgeConfig> = {
    allowGrabbing: true,
    allowOverlapSeparation: true,
    allowSelection: true,
    allowDeletion: false,
    allowEdit: false,
    // These should never be used and are set just to make Typescript happy
    allowEditSource: false,
    allowEditTarget: false,
}

/**
 * Get the config property for a node. If the node does not have the property, this falls back to the user node defaults, then the global node defaults.
 */
export function getNodeConfig<Prop extends keyof NodeConfig>(prop: Prop, node: NodeConfig, config: GrapherConfigSet): boolean {
    return node[prop] ?? config.nodeDefaults[prop] ?? globalNodeDefaults[prop]
}

/**
 * Get the config property for an edge. If the edge does not have the property, this falls back to the user edge defaults, then the global edge defaults.
 */
export function getEdgeConfig<Prop extends keyof EdgeConfig>(prop: Prop, edge: EdgeConfig, config: GrapherConfigSet): boolean {
    if (prop === "allowEditSource") return edge[prop] ?? config.edgeDefaults[prop] ?? getEdgeConfig("allowEdit", edge, config)
    if (prop === "allowEditTarget") return edge[prop] ?? config.edgeDefaults[prop] ?? getEdgeConfig("allowEdit", edge, config)
    return edge[prop] ?? config.edgeDefaults[prop] ?? globalEdgeDefaults[prop]
}

/**
 * Get the config property for a handle. If the handle does not have the property, this falls back to the corresponding node property, using
 * {@link getNodeConfig}. If the node does not have the property, this falls back to the user node defaults, then the global node defaults.
 */
export function getHandleConfig<Prop extends keyof NodeHandleConfig>(prop: Prop, handle: NodeHandleInfo, node: NodeConfig, config: GrapherConfigSet): boolean {
    if (handle[prop] != null) return handle[prop]!
    switch (prop) {
        case "allowGrabbing":
            return getNodeConfig("allowGrabbingHandles", node, config)
        case "allowNewEdges":
            return getNodeConfig("allowNewEdgesFromHandles", node, config)
        case "allowNewEdgeTarget":
            return getNodeConfig("allowNewEdgesFromHandles", node, config)
        default:
            errorCustom(`Internal error: unknown handle configuration property "${prop}"`)
            return false
    }
}