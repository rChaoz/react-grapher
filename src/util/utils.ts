import {warnCalcUnknownToken, warnUnknownCSSComputedValue} from "./log"

/**
 * Function that takes a computed CSS calc expression (i.e. composed of only percentage and pixel values, added or subtracted)
 * @param expr CSS calc expression: `<percentage>% +/- <pixel-value>px`
 * @return A pair of 2 numbers, as an array, first number being the percentage and second number the pixel value.
 * Note that the percentage is returned as-is: "50.3% + 5" will return [50, 5], not [.503, 5]
 * @see getComputedStyle
 */
export function splitCSSCalc(expr: string): [number, number] {
    // Splitting by whitespace is fine: in CSS, whitespace is mandatory before & after arithmetic '+'/'-'. Unary minus (or even plus)
    // can be handled by `Number()`, if present. Computed values may not contain multiplications or divisions, as they are considered basic
    // computations: only multiplying/dividing by 'number' (no unit) is allowed.
    const tokens = expr.trim().split(/\s/)
    let percentage = 0, absolute = 0, sign = 1
    for (const t of tokens) {
        if (t.length === 0) continue; // ignore consecutive spaces
        if (t.endsWith("px")) absolute += Number(t.substring(0, t.length - 2)) * sign
        else if (t.endsWith("%")) percentage += Number(t.substring(0, t.length - 1)) * sign
        else if (t === "+") sign = 1
        else if (t === "-") sign = -1
        else warnCalcUnknownToken(expr, t)
    }
    return [percentage, absolute]
}

/**
 * Resolve the argument of a computed CSS `calc()` function (i.e. composed of only percentage and pixel values, added or subtracted)
 * @param expr CSS calc expression: `<percentage>% +/- <pixel-value>px`
 * @param length length, in pixels, on which percentage values are based on
 * @see getComputedStyle
 */
export function resolveCSSCalc(expr: string, length: number) {
    const [percentage, absolute] = splitCSSCalc(expr)
    return percentage * length / 100 + absolute
}

/**
 * Convert a CSS computed value (`<number>px | <number>%`) to pixel value. For unknown values, this will return 0.
 * @param value CSS computed value
 * @param length If the value is or contains a percentage value, the length the percentage is based on (e.g. width/height of the parent for most
 * properties, or width/height of the element itself for border-radius).
  */
export function resolveValue(value: string, length: number): number {
    value = value.trim()
    // Resolve a percentage or pixel value to pixel value
    if (value.startsWith("calc(")) return resolveCSSCalc(value.substring(5, value.length - 1), length)
    else if (value.match(/^-?(\d+(\.\d+)?|\.\d+)?px$/)) return Number(value.slice(0, value.length - 2))
    else if (value.match(/^-?(\d+(\.\d+)?|\.\d+)?%$/)) return Number(value.slice(0, value.length - 1)) / 100 * length
    else {
        warnUnknownCSSComputedValue(value)
        return 0;
    }
}

/**
 * Convert 1 or two computed CSS values to pixel values (e.g. useful for border-top-left-radius, which may be 1 or 2 values)
 *
 * 'strValue' usually is a result of `getComputedStyle(...).someProperty`, and if of form:
 * ```
 * <value> | <value> <value>
 * ```
 * where <value> is:
 * ```
 * <number>px | <number>% | calc(<number>% +/- <number>px)
 * ```
 */
export function resolveValues(strValue: string, width: number, height: number): [number, number] {
    strValue = strValue.trim()
    const separator = strValue.startsWith("calc") ? strValue.indexOf(")", 5) + 1 : strValue.indexOf(" ")
    if (separator <= 0) return [resolveValue(strValue, width), resolveValue(strValue, height)]
    else return [resolveValue(strValue.substring(0, separator), width), resolveValue(strValue.substring(separator + 1), height)]
}

/**
 * Enlarge given rect (container) so that it contains another rect (child)
 */
export function expandRect(container: DOMRect, child: Pick<DOMRect, "x" | "y" | "width" | "height">) {
    // We don't use top/left/... properties on 'child' because either rect could be an SVGRect. Due to an issue in Typescript,
    // DOMRect type = SVGRect type, although SVGRect does *not* actually have top/left/... properties.
    // (https://github.com/microsoft/TypeScript-DOM-lib-generator/issues/974)
    if (child.x < container.x) {
        const delta = container.x - child.x
        container.x -= delta
        container.width += delta
    }
    if (child.y < container.y) {
        const delta = container.y - child.y
        container.y -= delta
        container.height += delta
    }
    if (child.x + child.width > container.right) container.width += child.x + child.width - container.right
    if (child.y + child.height > container.bottom) container.height += child.y + child.height - container.bottom
}

/**
 * Parse CSS value express as number|string. If number, "px" is appended to the value; otherwise the value itself is returned.
 * If value is null or the number 0, returns "0".
 */
export function convertToCSSLength(value: string | number | null | undefined) {
    if (value == null || value === 0) return "0"
    else if (typeof value === "number") return `${value}px`
    else return value
}

/**
 * Object used by {@link localMemo} to remember previous dependencies & memoized value.
 */
export type MemoObject<T> = {
    deps?: any[]
    oldValue?: T
}

/**
 * Memoize a value where a hook can't be used
 * @param factory Like {@link React.useMemo} factory param
 * @param deps Dependencies array, as value pairs: each pair consists of an old and new value that will be compared by reference
 * @param memoObject A memo object that will be used and modified by this function for memoization. Should be initialized to `{}`.
 * Note: the object must always be the same every time this function is called, for this function to work correctly.
 */
export function localMemo<T>(factory: () => T, deps: any[], memoObject: MemoObject<T>): T {
    if (memoObject.deps == null || memoObject.deps.length != deps.length) {
        memoObject.deps = deps
        return memoObject.oldValue = factory()
    }
    // Compare old & new dependencies
    for (let i = 0; i < deps.length; ++i) if (!Object.is(deps[i], memoObject.deps[i])) {
        memoObject.deps = deps
        return memoObject.oldValue = factory()
    }

    return memoObject.oldValue!
}

/**
 * Test whether an object has a property. The point of this function over `"prop" in object` is to provide return the same object,
 * but with the correct type inferred (i.e. same type but with an additional property 'prop').
 */
export function hasProperty<O extends object, P extends string>(object: O, prop: P): object is O & Record<P, unknown> {
    return prop in object
}