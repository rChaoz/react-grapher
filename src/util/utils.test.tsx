import React from "react"
import {describe, expect, it, jest, test} from "@jest/globals"
import styled from "@emotion/styled"

import {convertToCSSLength, deepEquals, expandRect, hasProperty, localMemo, resolveCSSCalc, resolveValue, resolveValues, splitCSSCalc} from "./utils"
import {render} from "@testing-library/react"

const Container = styled.div`
  position: relative;
  width: 1000px;
  height: 800px;
`

//noinspection CssReplaceWithShorthandSafely
const TestComponent = styled.div`
  position: absolute;

  width: 500px;
  height: 100px;

  /*
  = Position: =
  top:   398
  left:  150
  
  = Border radius: =
  property        x    y
  -------------  ---  ---
  top-left:      100   20
  top-right:     40    5
  bottom-right:  135   15
  bottom-left:   25    10
  */

  top: calc(50% - 2px);
  left: 15%;
  border-top-left-radius: 20%;
  border-top-right-radius: 40px calc(50% - 45px);
  border-bottom-right-radius: calc(25% + 10px) calc(30px - 15%);
  border-bottom-left-radius: calc(20px + 1%) 10%;
`

const consoleWarn = jest.spyOn(console, "warn").mockImplementation(() => {
    // empty
})

describe("splitCSSCalc", () => {
    test.each<[string, number, number]>([
        ["50% + 25px", 50, 25],
        ["30px", 0, 30],
        ["100px - 20%", -20, 100],
        ["25% - 5px", 25, -5],
    ])("works for '%s'", (expression, expectedPercentage, expectedPxValue) => {
        const [percentage, absolute] = splitCSSCalc(expression)
        expect(percentage).toBeCloseTo(expectedPercentage)
        expect(absolute).toBeCloseTo(expectedPxValue)
    })
})

describe("resolveCalc", () => {
    test.each<[number, string, number]>([
        [100, "50%", 50],
        [100, "20px", 20],
        [100, "10% + 25px", 35],
        [200, "15% -  4px", 26],
        [1000, "55px - 3%", 25],
        [350, "12px + 7%", 36.5],
    ])("using length=%d, calc(%s) = %s", (length, expression, expectedValue) => {
        expect(resolveCSSCalc(expression, length)).toBeCloseTo(expectedValue)
    })

    test("invalid calc expression", () => {
        // [500, "50px + invalid - 50vw - 1%", 45]
        // 'invalid' and '50vw' are invalid tokens
        consoleWarn.mockClear()
        expect(resolveCSSCalc("50px + invalid - 50vw - 1%", 500)).toBeCloseTo(45)
        expect(consoleWarn).toHaveBeenCalledTimes(2)
    })
})

// Render sample component to test against
const {container} = render(<Container><TestComponent/></Container>)
const elem = container.firstChild!.firstChild as HTMLElement

const style = window.getComputedStyle(elem)
// Width & height of element and parent element
// jsdom doesn't have a layout engine, so we need to set these manually (instead of using offsetWidth/offsetHeight)
const width = 500, height = 100
const parentWidth = 1000, parentHeight = 800


describe("resolvedValue", () => {
    it("works for simple property", () => {
        expect(resolveValue("50px", 0)).toBeCloseTo(50)
    })
    it("works for percentage property", () => {
        expect(resolveValue(style.top, parentHeight)).toBeCloseTo(398)
    })
    it("works for calc() property", () => {
        expect(resolveValue(style.left, parentWidth)).toBeCloseTo(150)
    })
    it("works for invalid property", () => {
        consoleWarn.mockClear()
        expect(resolveValue("invalid", 0)).toBe(0)
        expect(resolveValue("25vw", 0)).toBe(0)
        expect(consoleWarn).toHaveBeenCalledTimes(2)
    })
})

describe("resolvedValues", () => {
    it("works for single value property", () => {
        const borderTopLeftRadius = resolveValues(style.borderTopLeftRadius, width, height)
        expect(borderTopLeftRadius.length).toBe(2)
        expect(borderTopLeftRadius[0]).toBeCloseTo(100)
        expect(borderTopLeftRadius[1]).toBeCloseTo(20)
    })

    it("works for pixel value & calc property", () => {
        const borderTopRightRadius = resolveValues(style.borderTopRightRadius, width, height)
        expect(borderTopRightRadius.length).toBe(2)
        expect(borderTopRightRadius[0]).toBeCloseTo(40)
        expect(borderTopRightRadius[1]).toBeCloseTo(5)
    })

    it("works for calc & percentage value property", () => {
        const borderBottomLeftRadius = resolveValues(style.borderBottomLeftRadius, width, height)
        expect(borderBottomLeftRadius.length).toBe(2)
        expect(borderBottomLeftRadius[0]).toBeCloseTo(25)
        expect(borderBottomLeftRadius[1]).toBeCloseTo(10)
    })

    it("works for double calc property", () => {
        const borderBottomRightRadius = resolveValues(style.borderBottomRightRadius, width, height)
        expect(borderBottomRightRadius.length).toBe(2)
        expect(borderBottomRightRadius[0]).toBeCloseTo(135)
        expect(borderBottomRightRadius[1]).toBeCloseTo(15)
    })
})

// Because jsdom doesn't define it: https://github.com/jsdom/jsdom/issues/2716
class DOMRect {
    x: number
    y: number
    width: number
    height: number

    constructor(x = 0, y = 0, width = 0, height = 0) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }

    get left() {
        if (this.width < 0) return this.x + this.width
        else return this.x
    }

    get top() {
        if (this.height < 0) return this.y + this.height
        else return this.y
    }

    get right() {
        if (this.width > 0) return this.x + this.width
        else return this.x
    }

    get bottom() {
        if (this.height > 0) return this.y + this.height
        else return this.y
    }

    // noinspection JSUnusedGlobalSymbols
    toJSON() {
        return `{ "x": ${this.x}, "y": ${this.y}, "width": ${this.width}, "height": ${this.height} }`
    }
}

test("enlargeRect", () => {
    const rect = new DOMRect()

    expandRect(rect, new DOMRect(-10, -10, 20, 20))
    expect(rect).toMatchSnapshot()

    expandRect(rect, new DOMRect(50, 70))
    expect(rect).toMatchSnapshot()

    expandRect(rect, new DOMRect(-20, 0, 20, 100))
    expect(rect).toMatchSnapshot()

    expandRect(rect, new DOMRect(-80, 100, 50, 55))
    expect(rect).toMatchSnapshot()
})

describe("parseCSSStringOrNumber", () => {
    it("parses a number as a px string", () => {
        expect(convertToCSSLength(10)).toBe("10px");
    });

    it("returns a string unaltered", () => {
        expect(convertToCSSLength("5rem")).toBe("5rem");
        expect(convertToCSSLength("auto")).toBe("auto");
    });

    it("returns '0' for null, undefined or the number 0", () => {
        expect(convertToCSSLength(null)).toBe("0");
        expect(convertToCSSLength(undefined)).toBe("0");
        expect(convertToCSSLength(0)).toBe("0");
    });
})

describe('localMemo', () => {
    it('returns the memoized value when dependencies have not changed', () => {
        const factory = jest.fn(() => 100)
        const deps = ['a', 123, {obj: true}]
        const memoObject = {}

        expect(localMemo(factory, deps, memoObject)).toBe(100)

        // Change mock function implementation as it shouldn't be called again
        factory.mockImplementation(() => 0)

        expect(localMemo(factory, deps.slice(), memoObject)).toBe(100)

        expect(factory).toHaveBeenCalledTimes(1)
    })

    it('returns a new value when dependencies have changed', () => {
        const factory = jest.fn(() => "value 1")
        const deps = ['a', 123, {obj: true}]
        const memoObject = {}

        expect(localMemo(factory, deps, memoObject)).toBe("value 1")

        // Change deps & mock implementation to test new value
        const newDeps = deps.slice()
        newDeps[1] = 125
        factory.mockImplementation(() => "value 2")

        expect(localMemo(factory, newDeps, memoObject)).toBe("value 2")

        expect(factory).toHaveBeenCalledTimes(2)
    })

    it('returns the same reference every time if deps is empty array', () => {
        const factory = jest.fn(() => ({a: 10, b: true, c: ["test"]}))
        const deps: any[] = []
        const memoObject = {}

        const result1 = localMemo(factory, deps, memoObject)
        const result2 = localMemo(factory, deps, memoObject)
        const result3 = localMemo(factory, deps, memoObject)

        expect(result1).toBe(result2)
        expect(result2).toBe(result3)
        expect(result1).toEqual({a: 10, b: true, c: ["test"]})
        expect(factory).toHaveBeenCalledTimes(1)
    })
})

test("hasProperty", () => {
    consoleWarn.mockClear()

    const object: { a: string } = {a: "hello", b: 123} as any
    expect(hasProperty(object, "b"))
    // Expect no compile error
    if (hasProperty(object, "b")) console.warn(object.b)
    expect(consoleWarn).toHaveBeenCalled()
})

describe("deepEquals", () => {
    it("returns true for the same object, function or simple type", () => {
        const obj = {a: 1}
        expect(deepEquals(obj, obj)).toBe(true)
        const str = "hello world", strArr = ["he", "ll", "o ", "world"]
        expect(deepEquals(str, strArr.join(""))).toBe(true)

        expect(deepEquals(4, +"4")).toBe(true)
        expect(deepEquals(null, null)).toBe(true)
        expect(deepEquals(undefined, undefined)).toBe(true)

        function f() {
            // empty
        }

        expect(deepEquals(f, f)).toBe(true)
    })

    it("returns false for null and undefined", () => {
        expect(deepEquals(null, undefined)).toBe(false)
        expect(deepEquals({}, undefined)).toBe(false)
        expect(deepEquals(undefined, {})).toBe(false)
    })

    it("returns false for objects with different types, constructors, prototypes or number of properties", () => {
        expect(deepEquals({}, [])).toBe(false)
        expect(deepEquals({}, "")).toBe(false)
        expect(deepEquals({}, 1)).toBe(false)
        const obj1 = {a: 1, b: "test", c: {}}
        const obj2 = {a: 1, b: "test"}
        expect(deepEquals(obj1, obj2)).toBe(false)
    })

    it("returns true for equivalent objects", () => {
        const obj1 = {a: 1}
        const obj2 = {a: {b: 2}}
        expect(deepEquals(obj1, {...obj1})).toBe(true)
        expect(deepEquals(obj2, {...obj2})).toBe(true)
    })

    it("returns false for nested objects with different properties", () => {
        expect(deepEquals({a: 1}, {a: 2})).toBe(false)
        expect(deepEquals({a: 1}, {b: 1})).toBe(false)
        expect(deepEquals({a: {b: 1}}, {a: {c: 2}})).toBe(false)
        expect(deepEquals({a: {b: 1}}, {a: {b: 2}})).toBe(false)
    })

    it("returns false for objects with different constructors or prototypes", () => {
        class MyClass1 {
            constructor(public a: number) {
            }
        }

        class MyClass2 {
            constructor(public a: number) {
            }
        }

        const obj1 = new MyClass1(1)
        const obj2 = new MyClass2(1)
        expect(deepEquals(obj1, obj2)).toBe(false)

        expect(deepEquals({}, Object.create(null))).toBe(false)
    })
})