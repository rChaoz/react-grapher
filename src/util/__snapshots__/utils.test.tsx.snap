// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enlargeRect 1`] = `"{ "x": -10, "y": -10, "width": 20, "height": 20 }"`;

exports[`enlargeRect 2`] = `"{ "x": -10, "y": -10, "width": 60, "height": 80 }"`;

exports[`enlargeRect 3`] = `"{ "x": -20, "y": -10, "width": 70, "height": 110 }"`;

exports[`enlargeRect 4`] = `"{ "x": -80, "y": -10, "width": 130, "height": 165 }"`;
